README.md
[![Actions Status](https://github.com/kingsimba/nc-runtime/workflows/CMake/badge.svg)](https://github.com/kingsimba/nc-runtime/actions)

NC-Runtime <!-- omit in toc -->
==========

- [Background](#background)
- [StringSlice](#stringslice)
  - [Implmentation of StringSlice](#implmentation-of-stringslice)
- [NcString](#ncstring)
- [Smart Pointer](#smart-pointer)
- [StackOrHeapAllocator](#stackorheapallocator)
- [ManualResetEvent](#manualresetevent)
- [Many more good stuffs](#many-more-good-stuffs)
- [Visual Studio Visualizer](#visual-studio-visualizer)
- [Design Notes](#design-notes)
  - [The internal of shared_ptr](#the-internal-of-shared_ptr)
  - [Try to use shared_ptr](#try-to-use-shared_ptr)
  - [Things I like & dislike about `std::shared_ptr<>`](#things-i-like--dislike-about-stdshared_ptr)
  - [Makes all memory in one piece with `sp`](#makes-all-memory-in-one-piece-with-sp)
  - [`auto` keyword](#auto-keyword)
  - [Log System](#log-system)
  - [Implement operator [] for `sp<NcArray>`](#implement-operator--for-spncarray)
  - [Closure(Lambda Functions)](#closurelambda-functions)

# Background

Unlike other modern programming languages, C++ lacks some important infrastructures.
Everyone have to invent their own wheel.

Since C++ 11, with `shared_ptr`, `thread`, `mutex` etc, the Standards bring some orders into the chaos. But I don't totally agree with their design.

This project contains the methodology & wheels which I found useful.
They may not be the best ideas. But I hope you may find some pieces useful.


# StringSlice

C++ did a very bad job with std::string. (If you don't believe me, does anyone remember how to split a `std::string` into pieces?) Because it's soo bad, almost every library (QT, abseil, OpenCV, MFC, etc) all invented their own ones.

In my opinion, string is so important and should have been hard-coded into the compiler! If so, literal string will be more efficient and there will no incentives for the chaos in almost every libraries.

Sadly, I'd have to invent my own wheel. I want a string class which is both intuitive and efficient.
`StringSlice` is created just for that. It's always on stack and cheap to copy. Most of all, it's intuitive to use. (The name 'slice' is borrowed from Golang's slice concept.)

```cpp
vector<StringSlice> piece = "hello_world"_s.split("_");
pieces[0]; // "hello"
pieces[1]; // "world"

"hello_world"_s.startsWith("hello"); // true
"hello_world"_s.endsWith("world"); // true
"HELLO"_s.equalsCaseIncensitive("hello"); // true

"hello"_s.slice(1, 3) //  "ell"
"hello"_s.sliceFrom(2) //  "llo"

"internationalization"_s.countSlice("tion") // 2
```

StringSlice doesn't have `operator+` (which is the root of many bad code). I added `join` and `format`. They are better and faster.

```cpp
std::vector<StringSlice> lines({"hello", "world", ""});
"\n"_s.join(lines) // "hello\nworld\n"

StringSlice::format("hello %s", "simba") // "hello world");
```

## Implmentation of StringSlice

Unlike C string, which requires a `'\0'` ending. Unlike `std::string` which is mutable and allocated from heap. `StringSlice` is a simple struct, lives on stack, and immutable. It only contains a start pointer and a length.

So creating a subslice is very cheap(just advance the pointer and reassign the length). 
Consider the following code, there is not a single `malloc()/free()` in it.

```cpp
StringSlice slice = "hello---world"_s;
StringSlice pieces[3];
int num = slice.splitWithLimit("---", pieces, countof(pieces));
ASSERT_EQ(num, 2);
EXPECT_TRUE(pieces[0] == "hello");
EXPECT_TRUE(pieces[1] == "world");
```



Most of the time, StringSlice manages the memory safely.
For example:

```cpp
auto str1 = "hello world"_s; // use static memory
auto str2 = StringSlice::format("hello %s", "world"); // allocate memory internally
```

But sometimes, you should be more careful.

```cpp
// `str1` takes ownership of the memory, and will call `free()` when it's not needed'
char* buffer = json_dumps(jsonNode);
auto str1 = StrinSlice::makeByTakingCString(buffer);

char buffer[MAX_PATH];
getcwd(buffer, MAX_PATH);
// Careful: str2 will be invalid when `buffer` goes out of scope
auto str2 = StringSlice::makeEphemeral(buffer);
// `StringSlice::make()` will make a copy of the memory. It's safer, albeit slower.
auto str3 = StringSlice::make(buffer);
```

# NcString

`NcString` is on heap. It's derived from `NcObject`.
For most cases, please use `StringSlice` instead of `NcString`.
Use `NcString` only when you need an `NcObject` derived object.
For example, when you want to put it into `NcArray<>` or `NcHashmap<>`.

```cpp
auto str1 = "hello---world"_str;
auto str2 = NcString::allocWithCString("hello---world");

// splitting a string will only create StringSlices. So it's very fast.
auto slices = str->split("---");
EXPECT_EQ(str->retainCount(), 3); // because each slice holds a reference to the string

// can conver to StringSlice
str->toSlice();
```

The functions in `NcString` is the same as in `StringSlice`, such as  `join`', `format`, `startsWith`, `endsWith`, `split`, etc.

# Smart Pointer

`std::shared_ptr` gives a acceptable direction in automating memory management. I too think they should go directly in to the compiler, like what ObjectC++ did.

Once again, disappointed, I created `sp` and `wp`. They are also compatible with `std::shared_ptr` and `std::weak_ptr`.

```cpp
// get() and use_count() works the same as std::shared_ptr
sp<NcData> data = NcData::allocWithContentsOfFile("hello_world.txt");
EXPECT_EQ(data->length(), 11);   // overload operator '->'
EXPECT_EQ(data.use_count(), 1);

// create sp
sp<NcString> str = NcString::allocWithCString("hello");

NcString* raw = str.get();         // strong -> raw
sp<NcString> s2 = retainToSp(raw); // raw -> strong

wp<NcString> w(str);           // strong -> weak
sp<NcString> s = w.lock();     // weak -> strong
if (s != nullptr)
{
    use(s);
}
```

There are some differences with `std::shared_ptr`:

1.  There is a base class `NcObject`. `sp<>` only works on it. While `std::shared_ptr<>` can work on any class.

    `NcObject` contains some common used methods. Such as `toString()`, `isKindOf()` and `equals()`.
    They can be used to make fancy things like:

    ```cpp
    // ""_str is short for NcString::allocWithCString(const char* str);
    auto arr = NcArray::allocWithObjects("hello"_str, "world"_str, NULL);

    // any object can be coverted to string
    auto str = NcString::format("array %@ with length %d", arr, arr->length());
    EXPECT_EQ(str, "array [\"hello\", \"world\"] with length 2");
    ```

    ```cpp
    // deep & generic comparison
    auto array1 = NcArray::allocWithObjects(obj1, obj2, NULL);
    auto array2 = NcArray::alloc();
    array2->push(obj1);
    array2->push(obj2);
    array1->equals(array2); // should be true
    ```

    ```cpp
    // type identification
    sp<NcObject> box = MyBox::alloc();
    EXPECT_TRUE(box->isKindOf<MyBox>());
    EXPECT_TRUE(box->isKindOf<NcObject>());
    EXPECT_FALSE(box->isKindOf<NcString>());
    ```
   
2.  There is a `operator[]` in `sp`. It makes indexing into array easier

    ```cpp
    auto arr = NcArray::allocWithObject(obj1, obj2, NULL);
    printf(arr[0]->toString());
    printf(arr[1]->toString());
    ```

3.  It supports manual reference counting through `retain()` and `release()`.
    though their uses should be restricted (like when working with 3rd-party C API).

    ```cpp
    auto userData = UserData::alloc();
    // retain() manually increase RC
    SomeOldSdk_setUserData((void*)retain(userData.get())); 
    SomeOldSdk_setDeleteCallback([](void* userdata) {
        UserData* data = (UserData*)userdata;
        release(data); // release() manually decrease RC
    });
    ```

4.  The memory layout is more efficient for some objects.

    Consider the following code:

    ```cpp
    auto str1 = std::make_shared<std::string>("hello");
    ```

    There are three memory blocks (all on heap) for str1:
        
    *   the control block used by `shared_ptr`
    *   the string object
    *   the contained string "hello"

    `std::make_shared<>` smartly combine the first two blocks in one `malloc()`. 
    But the contained string requires a second `malloc()`.

    With `NcString`, all 3 blocks are created in one `malloc()`.
    Please see the implementation for how it works.

    ```cpp
    auto str2 = NcString::allocWithCString("hello");
    ```

    The same optimization is used for `NcData`.

# StackOrHeapAllocator

In some functions, stack is enough for most cases. But occasionally, larger memory is required.

For example, in NcLog_write(), `char message[2048]' is very, very likely to be enough, but we can't count on it. StackOrHeapAllocator is invented exactly for situations like that.

```cpp
TEST(Stdlib, stackOrHeapAllocator) {
   u8* stack = (u8*)alloca(1024);
   StackOrHeapAllocator allocator(stack, 1024);
   EXPECT_EQ(allocator.allocArray<u8>(512) - stack, 0);
   EXPECT_EQ(allocator.allocArray<u8>(512) - stack, 512);
   EXPECT_GT(allocator.allocArray<u8>(1) - stack, 4096); // stack used up, so it's on heap

   // no leak
}
```

All allocated memory will be freed when `allocator` goes out of scope.

# ManualResetEvent

It simulate Win32 API `SetEvent()`. With C++11's `condition_variable` it's much easier.
I learned it from https://stackoverflow.com/questions/1501111/boost-equivalent-of-manualresetevent

```cpp
ManualResetEvent e;

std::thread t([&] {
   Thread::sleep(100);
   e.set();
});

// must wait until reset
TimeTick start = TimeTick::now();
e.wait();
TickDuration duration = TimeTick::now() - start;
EXPECT_GE(duration.ms(), 100);

t.join();

// not reset(). So wait on it will not block
start = TimeTick::now();
e.wait();
duration = TimeTick::now() - start;
EXPECT_LT(duration.ms(), 1);

// reset() again. Wait will block
e.reset();
start = TimeTick::now();
EXPECT_FALSE(e.waitWithTimeout(10));
duration = TimeTick::now() - start;
EXPECT_GE(duration.ms(), 10);
```

# Many more good stuffs

- Log System
- NcCache

# Visual Studio Visualizer

`nc-runtime.natvis` is used to improve the debugging experience in Visual Studio.

Visualizer are included for `NcArray`, `sp`, `NcString`, `StringSlice` etc

![](vs_visualizer/visualizer.png)

# Design Notes

To be honest, I fall behind in terms of modern C++. Because I held a biased and negative attitude 
towards it and didn't spend much time to learn it. 
And in the New Years Days of 2021, I decide to rediscover modern C++. 
And see how things can be different if I use C++ 14.

## The internal of shared_ptr

Before discussing the pros-and-cons of both implementations, I dive into the internal of `std::shared_ptr`

At first, I think `std::shared_ptr` will have to make one additional malloc() for the control block.
Until I read from https://www.nextptr.com/tutorial/ta1358374985/shared_ptr-basics-and-internals-with-examples:

> We mentioned above that the control block could either contain a pointer to
> the managed object or the object itself.
> The control block is dynamically allocated. Constructing the managed object in-place
> within the control block can avoid the two separate memory allocations for the object
> and the control block, resulting in an uncomplicated control block and better performance.
> The std::make_shared is a preferred way to construct a shared_ptr because it builds the
> managed object within the control block.

Later, I also learned there is `std::enabled_shared_from_this`. It can put the control block into the base object.

## Try to use shared_ptr

So I think `std::shared_ptr` might be good enough for my need. And write something out of it:

```cpp
class NcObject : public std::enable_shared_from_this<NcObject> {
};

class Time : public NcObject {
public:
  // I like allocWith() over constructors. Because it allows overloading
  static shared_ptr<Time> allocWithSeconds(int sec) { return std::make_shared<Time>(sec * 1000); }
  static shared_ptr<Time> allocWithMilliseconds(int ms) { return std::make_shared<Time>(ms); }

  // Problem: Constructors must be public. Because std::make_shared needs it.
  Time(int ms) : m_milliseconds(ms) {};

  sp<Time> self() {
     return static_pointer_case<Time>(shared_from_this());
  }

private:
  int m_milliseconds;
};

TEST(NcObject, basic) {
  auto t = Time::allocWithSeconds(10);

  // cast to base
  shared_ptr<NcObject> base = t;
  EXPECT_EQ(t.use_count(), 2);

  // cast to derived
  auto t2 = std::static_pointer_cast<Time>(base);
  EXPECT_EQ(t.use_count(), 3);
  EXPECT_TRUE(t.get() == t2.get()); // underlying the same object

  auto t3 = t->self();
  EXPECT_EQ(t.use_count(), 4);
}
```

`Time::self()` makes an important point to implement something like
`auto pieces = "hello-world"_str->split("-")`.
Each piece contains a strong reference to the original string.

## Things I like & dislike about `std::shared_ptr<>`

Like:

1. Strong/weak. It makes delegate design pattern more robust.
2. combine control block with the controlled object.

Dislike:

1. Constructors must be public. Some user may call delete or create the object on stack.
2. Convert from raw pointer back to `shared_ptr` is inconvenient.
3. Must call `.get()` to get raw pointer. (I changed my mind, I become neural on that.)
4. There are too many usage variants

   ```cpp
   void f(widget* );
   void f(widget& );
   void f(shared_ptr<widget> );
   void f(const shared_ptr<widget>& );
   ```

   [Sutter’s Mill](https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/) and [acel](https://stackoverflow.com/questions/3310737/should-we-pass-a-shared-ptr-by-reference-or-by-value) has some insights on that

## Makes all memory in one piece with `sp`

With self-implemented control-block in NcObject. I can do some deep optimization, like appending the text directly after the NcString object.
Thus creating the string with a single allocation.
See `allocWithCString::allocWithCString()`

## `auto` keyword


CHANGELOG.md
# Changelog

- 2022-08-09 Add `nc_swapEndian()`.
- 2022-08-05 Add `Quaternion::inverse()`, `Pose::inverse()`, Pose concatenation. Vector unary minus operator.
- 2022-08-05 Add ROS type conversion for `Vector3`, `Quaternion` and `Pose`.
- 2022-08-05 Add `Pose3`.
- 2022-08-03 Add `LineParser`.
- 2022-08-01 Add `NC_ASSERT_NEAR()`
- 2022-07-27 Add `Math_polygonArea()`.
- 2022-07-20 Add `Vector3::zero()`, `Vector3::fill()`, `Quaternion::zero()`, `Quaternion::identity()`.
- 2022-06-20 Fix `sp<>` self-assignment.
- 2022-06-16 Add `Task`.
- 2022-05-18 Add `raytraceLineF()`.
- 2022-05-05 Add `Thread::setName()`.
- 2022-04-28 Implement `FileOpenFlag::createAlways` and `FileOpenFlag::openExisting`.
- 2022-04-28 Add `NcPolyline::calculateCentroid()`.
- 2022-04-25 Add `LineSimplifier`.
- 2022-04-24 Add `StringSlice::makeWithArgs()`.
- 2022-04-18 Add DBSCAN clustering algorithm: `nc::Dbscan`.
- 2022-03-23 Add `TimeDuration::makeWithString()`.
- 2022-03-16 Add `Resampler`.
- 2022-03-15 Use template for Vector3 and Quaternion.
- 2022-03-15 Add `double` version of some functions - `nc_wrapToTwoPi()`, `Math_mod()`, `Math_turnAngle()`.
- 2022-03-13 Add operator `==` and `!=` for `Pose2`.
- 2022-03-13 Add `JsonNode::array()` and `JsonNode:append()`.
- 2022-03-01 Add `Math_rectPointDistance()`.
- 2022-02-28 Add `FixedCircularArray`.
- 2022-02-22 Add `ImageU8::allocWithData()`.
- 2022-01-19 Add `RectF::isNegtiveMinimum(), RectF::expand(float width, float height)`.
- 2022-01-13 Add `StableRate` for loop rate control.
- 2022-01-07 Add `TimeDuration`. Add a lot of operator overloads for `TimeDuration` and `TimeTick`.
- 2021-12-31 Add default constructor for `Some<T>`.
- 2021-12-10 Improve JsonNode. Remove `Some<JsonNode>` etc. |api-break|

  - Remove `Some<JsonNode>`. JsonNode itself can be compared with `nullptr`.
  - Add `JsonNode::nullObject()`. It is a valid json node and totally different from `nullptr`.
  - Fix bug in JsonNode::dump().

- 2021-12-09 Add `StringSlice::cstr()`.

  `cstr()` may make a copy of the memory if it's not '\0' ended.
  Deprecate StringSlice::toCString().

- 2021-12-08 RCSS-813 related:

  - Add `ImagePainter::line()`.
  - Add parameter `CustomInflationFunc func` to constructor of `CostmapInflator`.
  - Add `CostmapInflator::setSeed()`.

- 2021-12-06 Add Base64.
- 2021-11-25 Add StringSlice::makeWithContentsOfFile().
- 2021-11-25 Fix bug in StringSlice::trim().
- 2021-11-24 Add StringSlice::trim(), trimStart(), trimEnd().
- 2021-11-24 Support conversion between std::string and StringSlice.
- 2021-11-20 Major improves on `StringSlice`, `sp<T>`

  - StringSlice become the major player, instead of NcString.
  - Add `join`, `format`, etc in StringSlice.
  - Disallow implict conversion from raw pointer to `sp<T>`. Use `retainAsSp()`.

- 2021-11-15 Support JSON modification with`JsonNode::remove()`, `JsonNode::add()`, `JsonNode::object()`, `JsonNode::integer()`, `JsonNode::real()` etc.
- 2021-11-15 Improve `NcObject::alloc()`.
- 2021-11-08 Add `NC_DEPRECATED`
- 2021-11-05 Add `NcObject::alloc<T>()`. Avoid using of new/delete.
- 2021-11-05 Add StateMachine.
- 2021-10-26 Add `Math_segmentPointDistance()`.
- 2021-10-22 Fix a bug in `RectF::setAsNegativeMinimum()`.
- 2021-10-18 Add `Math_linePointDistance()` and `Math_segmentsIntersect()`.
- 2021-10-14 Add `Vector2i` and `Math_clipSegmentByRect()`.
- 2021-10-12 Add `Rect::setAsNegativeMinimum()`, `Rect::combinePoint()` and `Rect::combineRect()`.
- 2021-10-05 Add `Math_pointInPolygon()` and `NcPolyline::testPoint()`.
- 2021-09-28 Fixed bug in JsonNode::asFloat().
- 2021-09-26 Add `NcPolyline` to support polyline and polygon.
- 2021-09-15 Add `CostmapInflator::inflateInplace()`.
- 2021-09-13 Add NcString::join() and NcString::allocByJoiningStrings().
- 2021-09-03 Add Mutex, LockGuard. Use GUARDED_BY() to detect thread error with clang.
- 2021-09-03 Add StringSlice::equalsCaseInsensitive().
- 2021-09-03 Add JsonNode::asBool(), JsonNode::isTrue(), JsonNode::isFalse().
- 2021-08-18 Add nc_ntohl() nc_ntohs(), nc_flipU16(), nc_flipU32() for endian swapping.
- 2021-08-18 `ResetableEvent` supports auto-reset.
- 2021-08-04 Add `Vector3` and some common operations.
- 2021-07-14 Add `JsonSettingLoader`
- 2021-07-14 Add "nc_argparse.h"
- 2021-07-13 Add `Math_turnAngle()`, `Math_wrapTwoPi()`
- 2021-07-02 Add `nc_abs()`, `nc_min()`, `nc_max()`, `nc_clamp()`
- 2021-07-02 Add `Quaternion` and `QuaterionD`.
- 2021-06-30 Add `nc_isnan()`.
- 2021-06-30 Add `nc_radianToDegree()` and `nc_degreeToRadian()`.
- 2021-06-30 Add single & double precision PI. NC_PI, NC_PI_2, NC_PI_D, NC_PI_2_D etc.
- 2021-06-22 Add NcImage, NcImageU8 and CostmapInflator.
- 2021-06-22 Add `StandardFreer`, `nc_copyArray()`.
- 2021-03-29 Add NC_ASSERT()
- 2021-03-24 Add `Some` and `noValue`.
- 2021-03-24 Add NcFile & NcData.
- 2021-03-05 Fix bug in copying of wp.
- 2021-03-05 Fix bug in global strings.
- 2021-02-28 Add wp, equivalent of std::weak_ptr.
- 2021-02-18 Add Thread::sleep(). Make TimeTick a class.
- 2021-02-18 Change the implementation of `sp<NcArray>`'s operator[].
- 2021-02-18 Add `synchroized()` and `ManualResetEvent`.
- 2021-02-17 Add NcArray::findWithCondition(), indexOfObject(), indexOfObjectWithCondition()
- 2021-02-17 Update Visual Studio Visualizer.
- 2021-02-16 Add NcString::format(). StringSlice::tokenize().
- 2021-02-16 Add log system. Add StackOrHeapAllocator.
- 2021-02-16 Add StringSlice::countSlice()
- 2021-02-16 "hello"\_s will create a StringSlice. Add startsWith() and endsWith().
- 2021-02-16 Add Math_hashU32(), Math_hashU64() and Math_hashSizeT()
- 2021-02-16 Support string literal. "hello"\_str will create a NcString object.
- 2021-02-16 Optimize NcString. Use a single allocation to create the string.
- 2021-02-16 Add subsliceFrom(), subsliceInRrange(), splitWithLimit(), replaceInRange() in StringSlice & NcString.
- 2021-02-15 Add NcObject::isKindOf()


src/nc_math.cpp
#include "stdafx_nc_runtime.h"
#include "nc_runtime/nc_math.h"

using namespace nc;

u32 Math_hashString(const char* str)
{
    u32 hash = 5381;
    int c;

    while ((c = *(u8*)(str)++) != 0)
        hash = ((hash << 5) + hash) + c; /* hash * 33 + c */

    return hash;
}

u32 Math_hashU32(u32 x)
{
    x = ((x >> 16) ^ x) * 0x45d9f3b;
    x = ((x >> 16) ^ x) * 0x45d9f3b;
    x = (x >> 16) ^ x;
    return x;
}

u64 Math_hashU64(u64 x)
{
    x = (x ^ (x >> 30)) * UINT64_C(0xbf58476d1ce4e5b9);
    x = (x ^ (x >> 27)) * UINT64_C(0x94d049bb133111eb);
    x = x ^ (x >> 31);
    return x;
}

size_t Math_hashSizeT(size_t o)
{
#if defined(NC_ENVIRONMENT_64)
    return Math_hashU64(o);
#else
    return Math_hashU32(o);
#endif
}

// Floating-point modulo
// The result (the remainder) has same sign as the divisor.
// Similar to matlab's mod(); Not similar to fmod() -   Mod(-3,4)= 1   fmod(-3,4)= -3
template <typename T>
static T _mod(T x, T y)
{
    static_assert(!std::numeric_limits<T>::is_exact, "Mod: floating-point type expected");

    if (0. == y)
        return x;

    T m = x - y * std::floor(x / y);

    // handle boundary cases resulted from floating-point cut off:

    if (y > 0) // modulo range: [0..y)
    {
        if (m >= y) // Mod(-1e-16             , 360.    ): m= 360.
            return 0;

        if (m < 0)
        {
            if (y + m == y)
                return 0; // just in case...
            else
                return y + m; // Mod(106.81415022205296 , _TWO_PI ): m= -1.421e-14
        }
    }
    else // modulo range: (y..0]
    {
        if (m <= y) // Mod(1e-16              , -360.   ): m= -360.
            return 0;

        if (m > 0)
        {
            if (y + m == y)
                return 0; // just in case...
            else
                return y + m; // Mod(-106.81415022205296, -_TWO_PI): m= 1.421e-14
        }
    }

    return m;
}

float Math_mod(float x, float y)
{
    return _mod(x, y);
}

double Math_mod(double x, double y)
{
    return _mod(x, y);
}

float Math_wrapTwoPi(float angle)
{
    return _mod(angle, NC_TWO_PI);
}

double Math_wrapTwoPi(double angle)
{
    return _mod(angle, NC_TWO_PI_D);
}

float Math_turnAngle(float from, float to)
{
    from = Math_wrapTwoPi(from);
    to = Math_wrapTwoPi(to);
    float diff = to - from;
    if (diff > NC_PI)
        diff -= NC_TWO_PI;
    else if (diff < -NC_PI)
        diff += NC_TWO_PI;
    return diff;
}

double Math_turnAngle(double from, double to)
{
    from = Math_wrapTwoPi(from);
    to = Math_wrapTwoPi(to);
    double diff = to - from;
    if (diff > NC_PI_D)
        diff -= NC_TWO_PI_D;
    else if (diff < -NC_PI_D)
        diff += NC_TWO_PI_D;
    return diff;
}

/* ======= Crossings Multiply algorithm =================================== */
/* http://www.realtimerendering.com/resources/GraphicsGems//gemsiv/ptpoly_haines/ptinpoly.cpp */

/*
 * This version is usually somewhat faster than the original published in
 * Graphics Gems IV; by turning the division for testing the X axis crossing
 * into a tricky multiplication test this part of the test became faster,
 * which had the additional effect of making the test for "both to left or
 * both to right" a bit slower for triangles than simply computing the
 * intersection each time.  The main increase is in triangle testing speed,
 * which was about 15% faster; all other polygon complexities were pretty much
 * the same as before.  On machines where division is very expensive (not the
 * case on the HP 9000 series on which I tested) this test should be much
 * faster overall than the old code.  Your mileage may (in fact, will) vary,
 * depending on the machine and the test data, but in general I believe this
 * code is both shorter and faster.  This test was inspired by unpublished
 * Graphics Gems submitted by Joseph Samosky and Mark Haigh-Hutchinson.
 * Related work by Samosky is in:
 *
 * Samosky, Joseph, "SectionView: A system for interactively specifying and
 * visualizing sections through three-dimensional medical image data",
 * M.S. Thesis, Department of Electrical Engineering and Computer Science,
 * Massachusetts Institute of Technology, 1993.
 *
 */

/* Shoot a test ray along +X axis.  The strategy is to compare vertex Y values
 * to the testing point's Y and quickly discard edges which are entirely to one
 * side of the test ray.  Note that CONVEX and WINDING code can be added as
 * for the CrossingsTest() code; it is left out here for clarity.
 *
 * Input 2D polygon _pgon_ with _numverts_ number of vertices and test point
 * _point_, returns 1 if inside, 0 if outside.
 */
bool Math_pointInPolygon(const Vector2* pgon, size_t numverts, Vector2 pt)
{
    if (numverts <= 2)
        return false;

    bool inside = false;
    bool yflag0, yflag1;

    const Vector2* vtx0 = &pgon[numverts - 1];
    const Vector2* vtx1 = &pgon[0];

    /* get test bit for above/below X axis */
    yflag0 = (vtx0->y >= pt.y);

    for (int j = (int)numverts + 1; --j;)
    {
        yflag1 = (vtx1->y >= pt.y);
        /* Check if endpoints straddle (are on opposite sides) of X axis
         * (i.e. the Y's differ); if so, +X ray could intersect this edge.
         * The old test also checked whether the endpoints are both to the
         * right or to the left of the test point.  However, given the faster
         * intersection point computation used below, this test was found to
         * be a break-even proposition for most polygons and a loser for
         * triangles (where 50% or more of the edges which survive this test
         * will cross quadrants and so have to have the X intersection computed
         * anyway).  I credit Joseph Samosky with inspiring me to try dropping
         * the "both left or both right" part of my code.
         */
        if (yflag0 != yflag1)
        {
            /* Check intersection of pgon segment with +X ray.
             * Note if >= point's X; if so, the ray hits it.
             * The division operation is avoided for the ">=" test by checking
             * the sign of the first vertex wrto the test point; idea inspired
             * by Joseph Samosky's and Mark Haigh-Hutchinson's different
             * polygon inclusion tests.
             */
            if (((vtx1->y - pt.y) * (vtx0->x - vtx1->x) >= (vtx1->x - pt.x) * (vtx0->y - vtx1->y)) == yflag1)
            {
                inside = !inside;
            }
        }

        /* Move to the next pair of vertices, retaining info as possible. */
        yflag0 = yflag1;
        vtx0 = vtx1;
        vtx1++;
    }

    return inside;
}

float Math_polygonArea(const nc::Vector2* points, size_t count)
{
    float twiceArea = 0.0f;
    for (size_t i = 0; i < count; i++)
    {
        int n = (i + 1) % count;
        twiceArea += points[i].x * points[n].y - points[n].x * points[i].y;
    }

    return twiceArea / 2;
}

float Math_linePointDistance(nc::Vector2 pt, nc::Vector2 p1, nc::Vector2 p2, nc::Vector2* projectionOut)
{
    Vector2 v = p2 - p1;
    Vector2 w = pt - p1;

    float c1 = Vector2_dot(w, v);
    float c2 = Vector2_dot(v, v);
    float b = c1 / c2;
    Vector2 pb = p1 + v * b;
    Vector2 tmp = pt - pb;
    float d = tmp.length();

    if (projectionOut != NULL)
        *projectionOut = pb;

    return d;
}

float Math_segmentPointDistance(nc::Vector2 pt, nc::Vector2 p1, nc::Vector2 p2, nc::Vector2* nearestPtOut)
{
    Vector2 v = p2 - p1;
    Vector2 w = pt - p1;

    float c1 = Vector2_dot(w, v);
    float c2 = Vector2_dot(v, v);
    float b = nc_clamp(c1 / c2, 0.0f, 1.0f);
    Vector2 pb = p1 + v * b;
    Vector2 tmp = pt - pb;
    float d = tmp.length();

    if (nearestPtOut != NULL)
        *nearestPtOut = pb;

    return d;
}

bool Math_segmentsIntersect(nc::Vector2 a1, nc::Vector2 a2, nc::Vector2 b1, nc::Vector2 b2)
{
    if (std::max(a1.x, a2.x) < std::min(b1.x, b2.x) || std::max(b1.x, b2.x) < std::min(a1.x, a2.x)
        || std::max(a1.y, a2.y) < std::min(b1.y, b2.y) || std::max(b1.y, b2.y) < std::min(a1.y, a2.y))
        return false;

    return Vector2_cross(a2 - a1, b1 - a1) * Vector2_cross(a2 - a1, b2 - a1) <= 0
           && Vector2_cross(b2 - b1, a1 - b1) * Vector2_cross(b2 - b1, a2 - b1) <= 0;
}

namespace
{
enum class RectRegionCode
{
    inside = 0,
    left = 1,
    right = 2,
    top = 4,
    bottom = 8
};

static forceinline RectRegionCode operator|(RectRegionCode r, RectRegionCode l)
{
    return (RectRegionCode)((int)r | (int)l);
}

static forceinline RectRegionCode& operator|=(RectRegionCode& r, RectRegionCode l)
{
    return r = r | l;
}

static forceinline bool operator&(RectRegionCode r, RectRegionCode l)
{
    return ((int)r & (int)l) != 0;
}

class BoundaryFloat
{
public:
    static forceinline bool greater(float l, float r) { return l > r; }
    static forceinline float value() { return 0; }
};

class BoundaryInt
{
public:
    static forceinline bool greater(int l, int r) { return l >= r; }
    static forceinline int value() { return 1; }
};
} // namespace

template <typename RECT, typename VECTOR, typename Boundary>
static RectRegionCode _calcRectRegionCode(RECT rect, VECTOR pt)
{
    RectRegionCode code = RectRegionCode::inside;

    if (pt.x < rect.left)
        code = RectRegionCode::left;
    else if (Boundary::greater(pt.x, rect.right))
        code = RectRegionCode::right;

    if (pt.y < rect.top)
        code |= RectRegionCode::top;
    else if (Boundary::greater(pt.y, rect.bottom))
        code |= RectRegionCode::bottom;

    return code;
}

/**
 * @sa https://www.geeksforgeeks.org/line-clipping-set-1-cohen-sutherland-algorithm/
 */
template <typename RECT, typename VECTOR, typename Boundary>
static bool _clipLineByRect(RECT rect, VECTOR* p1, VECTOR* p2)
{
    RectRegionCode code1 = _calcRectRegionCode<RECT, VECTOR, Boundary>(rect, *p1);
    RectRegionCode code2 = _calcRectRegionCode<RECT, VECTOR, Boundary>(rect, *p2);

    for (;;)
    {
        if (code1 == RectRegionCode::inside && code2 == RectRegionCode::inside)
            return true;

        if (code1 & code2)
            return false;

        RectRegionCode codeOut = (int)code1 ? code1 : code2;
        VECTOR p;

        /*
         *  Find intersection point, using formulas:
         *  y = y1 + slope * (x - x1)
         *  x = x1 + (1 / slope) * (y - y1)
         */
        if (codeOut & RectRegionCode::left)
        {
            /* point is to the left of rectangle */
            p.y = p1->y + (p2->y - p1->y) * (rect.left - p1->x) / (p2->x - p1->x);
            p.x = rect.left;
        }
        else if (codeOut & RectRegionCode::right)
        {
            /* point is to the right of rectangle */
            p.y = p1->y + (p2->y - p1->y) * (rect.right - Boundary::value() - p1->x) / (p2->x - p1->x);
            p.x = rect.right - Boundary::value();
        }
        else if (codeOut & RectRegionCode::top)
        {
            /* point is above the clip rectangle */
            p.x = p1->x + (p2->x - p1->x) * (rect.top - p1->y) / (p2->y - p1->y);
            p.y = rect.top;
        }
        else
        {
            NC_ASSERT(codeOut & RectRegionCode::bottom);
            /* point is below the rectangle */
            p.x = p1->x + (p2->x - p1->x) * (rect.bottom - Boundary::value() - p1->y) / (p2->y - p1->y);
            p.y = rect.bottom - Boundary::value();
        }

        if (codeOut == code1)
        {
            *p1 = p;
            code1 = _calcRectRegionCode<RECT, VECTOR, Boundary>(rect, *p1);
        }
        else
        {
            *p2 = p;
            code2 = _calcRectRegionCode<RECT, VECTOR, Boundary>(rect, *p2);
        }
    }
}

bool Math_clipSegmentByRect(RectF rect, Vector2* p1, Vector2* p2)
{
    return _clipLineByRect<RectF, Vector2, BoundaryFloat>(rect, p1, p2);
}

bool Math_clipSegmentByRectI(Rect rect, nc::Vector2i* p1, nc::Vector2i* p2)
{
    return _clipLineByRect<Rect, Vector2i, BoundaryInt>(rect, p1, p2);
}

float Math_rectPointDistance(nc::Vector2 pt, RectF rect, nc::Vector2* nearestPtOut)
{
    RectRegionCode code = _calcRectRegionCode<RectF, Vector2, BoundaryFloat>(rect, pt);
    if (code == RectRegionCode::inside)
    {
        if (nearestPtOut != nullptr)
            *nearestPtOut = pt;
        return 0.0f;
    }

    nc::Vector2 nearest;
    if (code & RectRegionCode::left)
    {
        nearest.x = rect.left;
    }
    else if (code & RectRegionCode::right)
    {
        nearest.x = rect.right;
    }
    else
    {
        nearest.x = pt.x;
    }
    if (code & RectRegionCode::bottom)
    {
        nearest.y = rect.bottom;
    }
    else if (code & RectRegionCode::top)
    {
        nearest.y = rect.top;
    }
    else
    {
        nearest.y = pt.y;
    }

    if (nearestPtOut != nullptr)
        *nearestPtOut = nearest;
    return hypotf(nearest.x - pt.x, nearest.y - pt.y);
}
